upf/upf.h

```diff
@@ -534,27 +534,39 @@ typedef enum
   _ (FLOWS_STITCHED_DIRTY_FIFOS, stitched_dirty_fifos, upf)                   \
   _ (TIMERS_MISSED, timers_missed, upf)
 
+enum
+{
+  URR_VOLUME_COUNTER_UL = 0,
+  URR_VOLUME_COUNTER_DL = 1,
+  URR_VOLUME_COUNTER_TOTAL = 2,
+  URR_VOLUME_N_COUNTERS = 3,
+};
+
 /* TODO: measure if more optimize cache line aware layout
  *       of the counters and quotas has any performance impcat */
-typedef struct
+typedef union
 {
-  u64 ul;
-  u64 dl;
-  u64 total;
-} urr_counter_t;
+  struct
+  {
+    u64 ul;
+    u64 dl;
+    u64 total;
+  };
+  u64 as_array[URR_VOLUME_N_COUNTERS];
+} urr_volume_counter_t;
```

Comment:
Not sure if you haev seen my comment, so let me repeat it as inline comment on the relevant lines.

I'm not convicend that this actually an improvement. The static function accessing the URRs will most likes be inlined (or can be force to be inline) in which case the compiler is highly likely to generate almost the same or at least very similar code for the old and the new version.
Changing the datastructur from an plain struct to an union where an array aliases the structure member makes this more error prone without adding any benefit.

Reply:
It's not improvement for compiler, but for reader. Mostly to get rid of `urr_incr_and_check` macro which does strange access to fields and make it clear which fields are accessed in urr_increment_and_check_counter. Also to get rid of offsetof in format.

Reply:
I see it exactly the other way arround, the change makes it IMHO harder for the reader to understand the code. 

Besides the style issue which could be down personal prefences, I still maintain that aliasing the field members through the union with the array is a bad idea. Constructs like that should only be used if there is absolutely not other god way to achive the same result.
If you insist on using an array, then use only the array and get rid of the struct.

The use of offsetof is valid and well known C syntax. I have no idea what would be a problem with that.

Reply:
Oh, okay. @demo-exe has similar position. So I will cut this commit to minimal change and prepare separate one with more types and use of arrays everywhere.